Expose seastar::sleep and clocks to Rust

patjed41 opened the pull request:
Fixes: #28
Depends on: #24, #27

I made a lot of changes to this PR. The old description made no sense after applying these changes so I decided to write a new description.

This PR does 2 things now:
- Exposes specializations of clocks along with its duration and type_point types (`clocks` module).
- Uses 2 of 3 exposed duration types to implement Rust specializations of `seastar::sleep` (`sleep` module).

The unused part of the `clocks` module is here just for organizational purposes. It is used in the timer PR (#32).

# Clocks

The `clocks` module introduces:
- `SteadyClock` − equivalent of `std::chrono::steady_clock`,
- `LowresClock` − equivalent of `seastar::lowres_clock`,
- `ManualClock` − equivalent of `seastar::manual_clock`.

All of these clocks are unit-like structs with an associated function `now`. Additionally, `ManualClock` has a function `advance`.

## Durations and instants

Apart from the clocks, duration and type_point (called *instant* in Rust) types related to them are exposed. All duration types and all instant types have the same definition so we use `duration_impl` and `instant_impl` macros to generate all of them. These types are very similar to `std::time::{Duration, Instant}`, but use `i64` underneath. The duration types have a little bit less methods than `std::time::Duration`. For example, they lack methods using floats and saturating methods. I didn't implement them for now because it would be time-consuming and I don't think it would bring much value to our project. Implementing more methods can be done in the future if really needed.

So why are these types worth the effort of creating them? We can use the fact that `std::chrono::steady_clock::duration` is almost always defined as `std::chrono::nanoseconds` (even though it's unspecified in the C++ standard) and `std::chrono::nanoseconds::rep` almost always equals `int64_t` (it's also unspecified). We can make our API work perfectly, that is without any unwanted conversions, if these types are defined as written above. Moreover, we can use `std::chrono::duration_cast` if  `std::chrono::steady_clock::duration` is not defined as `std::chrono::nanoseconds`. The only hopeless case is when `std::chrono::steady_clock::duration::rep` does not equal `int64_t`. In such a case static assertions defined in `clocks.cc` fail the compilation.

# Sleeps

Creating duration types simplified `seastar::sleep` specializations a lot. We don't need `sleep_secs, sleep_millis, ...` now. One sleep function per clock is enough:
- `steady_sleep` using `SteadyClock`,
- `lowres_sleep` using `LowresClock`.

patjed41 committed:
Add sleep function

This commit exposes `seastar::sleep` function that can be used with
all clocks implementing the `Clock` trait.

patjed41 committed:
Write tests for the sleep module

Lorak-mmk commented on seastar/src/config_and_start_seastar.cc:
```diff
@@ -54,4 +54,4 @@ int32_t run_int(app_template& app, int32_t argc, char** argv, IntFuture fut) {
 } // namespace config_and_start_seastar
-} // namespace seastar
\ No newline at end of file
+} // namespace seastar
```
This change is unrelated to this commit.
I don't have a problem with unrelated stylistic changed in a PR, but it should be in it's own commit

patjed41 replied:
Right, I moved this to a separate commit.

Lorak-mmk commented on seastar/src/sleep.hh:
```diff
@@ -0,0 +1,10 @@
+namespace sleep {
+
+} // namespace sleep
+} // namespace seastar_ffi
```
nit: no newline at end of file

Lorak-mmk commented on seastar/src/lib.rs:
```diff
@@ -7,6 +7,7 @@ mod cxx_async_local_future;
 
 mod config_and_start_seastar;
 mod preempt;
+#[allow(dead_code)]
```
Why is it needed?

patjed41 replied:
It turns out it isn't needed. I had *unused function* warning for all sleep functions. I don't have them now. I'm not sure what changed but it works now.

Lorak-mmk reviewed (changes requested):
Commit bdd4b5051acc4f1bd32f893b9e837df2c55501fd will have to be dropped after we merge #27

Please add some commit messages that explain shortly what is being changed and why.

patjed41 committed:
Create the sleep module

This commit creates files: `sleep.rs`, `sleep.hh`, `sleep.cc`. It also
sets up the sleep module in `lib.rs` and `build.rs`.

patjed41 commented:
Update:
- Made changes requested in a review.
- Added more verbose commit messages to commits where it seemed reasonable.

patjed41 commented:
> Commit [bdd4b50](https://github.com/zpp-2022-rust-seastar/seastar-rs/commit/bdd4b5051acc4f1bd32f893b9e837df2c55501fd) will have to be dropped after we merge #27

That was my plan. Honestly, I didn't know that to do with this. My PR depends on 2 PRs: #24, #27. I can rebase my branch on one of them, but what about the second one? I can do another rebase or use cherry-pick but no matter what I do, creating a commit with changes from the second branch seems to be necessary (if it isn't I'd like to know how to achieve this). So I have 2 options:
- Create a commit that *picks* the changes from the seconds branch – in this specific case it is the *Pick API safety module* commit. Then create another commit that uses these changes to do what is needed – in this case it is the *Ensure API safety of sleep funtions* commit. If I do it like this, as pointed out, the first of these commits will have to be dropped after merging the second PR.
- Create a single commit that does both of the things described above at once. If I do it like this, I won't have to change the history of this PR because all of the changes picked from the second branch will already be here after merging that branch and will disappear from the list of changes introduced by this single commit.

So, my question is, what is the better approach in the case like this one?

piodul commented:
> * Create a commit that _picks_ the changes from the seconds branch – in this specific case it is the _Pick API safety module_ commit. Then create another commit that uses these changes to do what is needed – in this case it is the _Ensure API safety of sleep funtions_ commit. If I do it like this, as pointed out, the first of these commits will have to be dropped after merging the second PR.
> * Create a single commit that does both of the things described above at once. If I do it like this, I won't have to change the history of this PR because all of the changes picked from the second branch will already be here after merging that branch and will disappear from the list of changes introduced by this single commit.
> 
> So, my question is, what is the better approach in the case like this one?

Cherry picking is definitely better. After the cherry-picked commit is merged, you can then rebase on main. Sometimes, git is smart enough to notice this and will automatically skip that commit during interactive rebase.

Instead of picking and then fixing things after the picked commit, you can start your PR from the #24's branch, cherry pick #27 and then write your implementation with the fixes applied already.

piodul commented on seastar/src/sleep.rs:
```diff
@@ -7,5 +7,38 @@ mod ffi {
+/// Uses steady clock defined in Seastar.
+pub async fn sleep_nanos(duration: i64) {
+    ffi::sleep_nanos(duration).await.unwrap();
+}
```
Please prefix those with `steady_` - just like you did with `lowres_` functions, so that it is clear which timer those functions are using.

patjed41 replied:
Done.

piodul commented on seastar/src/sleep.cc:
```diff
@@ -3,5 +3,21 @@
 namespace seastar_ffi {
 namespace sleep {
 
+VoidFuture sleep_secs(int32_t duration) {
+    co_await seastar::sleep(std::chrono::seconds(duration));
+}
+
+VoidFuture sleep_millis(int32_t duration) {
+    co_await seastar::sleep(std::chrono::milliseconds(duration));
+}
+
+VoidFuture sleep_micros(int32_t duration) {
+    co_await seastar::sleep(std::chrono::microseconds(duration));
+}
+
```
I don't think that we need those. A single `sleep_nanos` should be sufficient. 2^64 nanoseconds is about 500-600 years, so `sleep_nanos` should be sufficient (for some time, at least).

piodul replied:
Later, we can use `sleep_nanos` to build nicer, idiomatic abstractions - for example, introduce a `SteadyDuration` type which can be created from a number of seconds/milliseconds/etc. It would keep the duration as a number of nanoseconds so `sleep_nanos` could suffice.

patjed41 replied:
After introducing duration types this issue does not exist anymore.

patjed41 replied:
> Later, we can use `sleep_nanos` to build nicer, idiomatic abstractions - for example, introduce a `SteadyDuration` type which can be created from a number of seconds/milliseconds/etc. It would keep the duration as a number of nanoseconds so `sleep_nanos` could suffice.

I defined instant and duration types for all 3 clocks specializations.

There was a suggestion in a review of #32 that we could use `std::time::Duration`. It could be really easy to implement but I decided that creating our own duration types is worth the effort. I'll try to explain why:
- We would have to check if a given duration does not overflow `i64` in every function that uses duration. It is not a big problem currently since only sleeps and timer use durations but they might appear in other places (like `seastar::semaphore`) in the future if this project grows. Having to remember that checks are necessary wouldn't be nice.
- Something like *dont't use durations oveflowing `i64`* in documentation wouldn't look good. Readers would wonder why an overflowing type like this is even used.
- Different instant types using the same duration type would look very suspicious. In C++ the type_point types are dependant on the duration types.
- Overall, I think an API with our own duration types looks better. This is not a very signifacant difference but creating a type like that is not that hard and time-consuming so I don't see a reason why we wouldn't want it.

> I don't think that we need those. A single `sleep_nanos` should be sufficient. 2^64 nanoseconds is about 500-600 years, so `sleep_nanos` should be sufficient (for some time, at least).

After introducing duration types this issue does not exist anymore.

piodul commented on seastar/src/sleep.cc:
```diff
@@ -19,5 +19,13 @@ VoidFuture sleep_nanos(int64_t duration) {
     co_await seastar::sleep(std::chrono::nanoseconds(duration));
 }
 
+VoidFuture lowres_sleep_secs(int32_t duration) {
+    co_await seastar::sleep<seastar::lowres_clock>(std::chrono::seconds(duration));
+}
```
Same here - I think `lower_sleep_secs` is redundant.

patjed41 replied:
Defining duration types solved this.

piodul commented on seastar/src/sleep.rs:
```diff
@@ -59,3 +59,57 @@ pub async fn lowres_sleep_secs(duration: i32) {
+        let duration = 100;
+        sleep_millis(duration).await;
+        let elapsed = now.elapsed().unwrap().as_millis() as i32;
+        assert!(elapsed >= duration / 2 && elapsed < duration * 2);
```
I'm not sure about those. It's a bad practice in tests to expect for something to happen within a particular amout of time - the CI machines can sometimes be very slow and such assertions may lead to intermittent failures in CI.

IMO it should be fine to check that at least `duration` s/ms/ns passed, so maybe we could just stick to that?

patjed41 replied:
Right, I changed tests to check whether `elapsed >= duration`.

patjed41 committed:
Create the clocks module

This commit creates files: `clocks.rs`, `clocks.hh`, `clocks.cc`. It also
sets up the `clocks` module in `lib.rs` and `build.rs`.

patjed41 committed:
Add new line at the end of config_and_start_seastar.{hh, cc}

patjed41 committed:
Add instant type

This commit introduces the generic `Instant` type that is an equivalent of
`std::chrono::type_point`.

patjed41 committed:
Add duration type

This commit introduces the generic `Duration` type that is an equivalent of
`std::chrono::duration`.

patjed41 committed:
Expose clocks to Rust

This commit exports 3 clock specializations used by Seastar to Rust:
- `SteadyClock` - an equivalent of `std::chrono::steady_clock`,
- `LowresClock` - an equivalent of `seastar::lowres_clock`,
- `ManualClock` - an equivalent of `seastar::manual_clock`.

More precisely, `now` functions of all these clocks are exposed plus
`advance` function of `seastar::manual_clock`.

All clock specializations implement the Clock trait containing only `now`
function. Moreover, they implement the hidden trait
`clock_implementation::ClockImpl`. All clock specific ffi functions will
be placed in this trait.

C++ does not define the duration type used by `std::chrono::steady_clock`.
However, it is almost always defined as `std::chrono::nanoseconds` and
`std::chrono::nanoseconds::rep` (also not defined by the C++ standard) is
almost always defined as `int64_t`. We want to use this trend and create
our API in such a way that it works perfectly (that is without any
conversions) when these types are defined as written above.

If `std::chrono::steady_clock` does not use `std::chrono::nanoseconds`,
we can use `std::chrono::duration_cast`, so we can manage such a case.
However, if `std::chrono::steady_clock::duration::rep` does not equal
`int64_t`, we cannot avoid unwanted conversions. To make our API always
work properly or not compile, we use static assertions in `clocks.cc`.

patjed41 committed:
Write tests for the clocks module

patjed41 commented:
Update:
- Made changes requested in a review.
- Introduced the `clocks` module (see a new description of this PR). These changes make part of #32 that exposes clocks outdated. I'll be working on fixing that PR from now.

Lorak-mmk commented on seastar/src/clocks.rs:
```diff
@@ -6,3 +8,135 @@ mod ffi {
+// - `Duration` - name of the corresponding duration type,
+// - `Clock` - name of the corresponding clock.
+macro_rules! instant_impl {
+    ($InstantName:ident, $Instant:ty, $Duration:ty, $Clock:ty) => {
```
Do we need both `InstantName` and `Instant`? Can't we use one of them in all the places?

patjed41 replied:
The problem was that with `$Instant:ty`
```rust
pub struct $Instant {
    pub(crate) nanos: i64,
}
```
didn't compile and I had to use `ident` instead of `ty` there.

I tried the other way − `ident` in all places − and it works. So I changed this macro to use `$Instant:ident, $Duration:ty, $Clock:ty`.

Lorak-mmk commented on seastar/src/clocks.rs:
```diff
@@ -6,3 +8,135 @@ mod ffi {
+            pub fn duration_since(&self, other: Self) -> $Duration {
+                match self.nanos.checked_sub(other.nanos) {
+                    Some(nanos) => <$Duration>::from_nanos(nanos),
+                    None => panic!(),
```
All of the functions in this impl return options, but this one panics. Why the inconsistency?

patjed41 replied:
I based my instant types on `std::timer::Instant`. This type also has *checked* and *unchecked* versions of `duration_since`:
- [std::timer::Instant::duration_since](https://doc.rust-lang.org/std/time/struct.Instant.html#method.duration_since),
- [std::timer::Instant::checked_duration_since](https://doc.rust-lang.org/std/time/struct.Instant.html#method.checked_duration_since).

What's quite surprising is that `std::timer::Instant::duration_since` panics or saturates depending on the Rust version. I decided to consistently panic (instead of saturating) in all methods of my instant and duration types.

Implementing `duration_since` probably isn't necessary. I was simply implementing all methods that `std::timer::Instant` implements. So, in conclusion, my instant types are like `std::timer::Instant` with `std::chrono::type_point`'s features:
- they use a signed integer underneath,
- they implement `duration_since_epoch` (I called it like that instead of `time_since_epoch` because the rest of the methods use *duration*).

Lorak-mmk commented on seastar/src/clocks.rs:
```diff
@@ -6,3 +8,135 @@ mod ffi {
+                };
+            }
+        }
+
+        impl Sub<$Duration> for $Instant {
+            type Output = Self;
+
+            /// # Panics
+            /// Panics if an overflow in the underyling data structure happens.
+            fn sub(self, duration: $Duration) -> Self {
+                match self.nanos.checked_sub(duration.nanos) {
+                    Some(nanos) => Self { nanos },
+                    None => panic!(),
+                }
+            }
+        }
+
+        impl SubAssign<$Duration> for $Instant {
+            /// # Panics
+            /// Panics if an overflow in the underyling data structure happens.
+            fn sub_assign(&mut self, duration: $Duration) {
+                match self.nanos.checked_sub(duration.nanos) {
+                    Some(nanos) => self.nanos = nanos,
+                    None => panic!(),
+                };
+            }
+        }
+
+        impl Sub<$Instant> for $Instant {
+            type Output = $Duration;
+
+            /// Returns the amount of time elapsed from another instant to this one.
+            /// If `other` is later that `self` the returned value is negative.
+            ///
+            /// # Panics
+            /// Panics if an overflow in the underyling data structure happens.
+            fn sub(self, other: $Instant) -> $Duration {
+                self.duration_since(other)
+            }
+        }
```
`AddAssing` and `SubAssign` have to panic on overflow, but `Add` and `Sub` don't. Considering that most of the functions in normal `impl` don't panic, maybe it's a better approach to make `Add` and `Sub` not panicking? Not sure what's the best approach here - it's a question of reliability vs ease of use. @piodul ?

piodul replied:
Let's stick to the conventions established in the Rust standard library. I can see that adding/subtracting instants/durations do not return options, so we should panic.

Lorak-mmk commented on seastar/src/clocks.rs:
```diff
@@ -140,3 +140,244 @@ macro_rules! instant_impl {
+                Self { nanos: -self.nanos }
+            }
+        }
+    };
```
Here it's the same consideration: should non-Assing version panic? @piodul

Lorak-mmk commented on seastar/src/clocks.rs:
```diff
@@ -140,3 +140,244 @@ macro_rules! instant_impl {
+                if self.nanos == i64::MAX {
+                    panic!();
+                }
+                Self { nanos: -self.nanos }
```
It doesn't look correct. Did you min `i64::MIN`? To be sure it works you could do 0.checked_sub(self.nanos)

patjed41 replied:
Yes, it was incorrect. Using `0.checked_sub(self.nanos)` is a really good idea. For some reason I had to write `0_i64.checked_sub(self.nanos)` to make it compile.

Lorak-mmk commented on seastar/src/clocks.rs:
```diff
@@ -140,3 +140,244 @@ macro_rules! instant_impl {
+            /// # Panics
+            /// Panics if result overflows.
+            fn add(self, rhs: Self) -> Self {
+                match self.nanos.checked_add(rhs.nanos) {
```
Why not `self.checked_add`?
Then it could be just `self.checked_add(&rhs).unwrap()`

The same applies to all the other impls.

patjed41 replied:
Good idea, simplified the code a lot.

Lorak-mmk replied:
I see you fixed that in duration impls, but not in instant impls, is there a reason for that?

Lorak-mmk replied:
By the way, after applying this suggestion, all the arithmetic impls `impl Add, impl Mul, impl SubAssign etc` for both instants and durations will no longer need to know about fields of the structures - they will just call methods. 
I'm wondering if we can't create better API by:
1. Creating a trait for instant and a trait for duration - with the methods that are currently in the main impl block.
2. Write arithmetic impls once for those traits.
3. Macro will then just need to generate the structure itself and an implementation of this new trait.

I see some advantages of this approach:
- The biggest one: user will be able to write methods that are generic over instants / durations.
- Less source code after macro expansion (slightly faster compilation)
- Not sure about amount of generated code: possibly there is less of it, as rustc won't have to generate code for unused impl specializations?

patjed41 replied:
> I see you fixed that in duration impls, but not in instant impls, is there a reason for that?

I forgot about that. It's fixed now.

After thinking about this new API for a moment I don't see a reason why this wouldn't work. Couldn't we also create a `Clock` trait that would implement only `now` function? Manual clock is a little different so maybe it's not the best idea.

I'm not sure if we want to implement this API and rewriting our current API seems to be a bit time-consuming so I've refrained from doing it for now. Maybe we will discuss this during tomorrow's meeting.

Lorak-mmk reviewed (changes requested):
Left some comments, overall looks pretty good.

It would be nice if there were some comments in the tests explaining what is the test testing etc - for example, line 534, `assert!(catch_unwind(|| i3.duration_since(i2)).is_err());`: it's not obvious on the first glance why this should panic, so a comment explaining that i3 is i64::MAX, i2 is less than zero, thus difference exceeds i64::MAX would be helpful. 
Such things may be obvious to the author right now, but it's hard to debug complicated tests later on when you can't understand them.

piodul commented on seastar/src/clocks.rs:
```diff
@@ -6,3 +8,134 @@ mod ffi {
+                }
+            }
+
+            /// Returns `Some(t)` where `t` ii equal to `self - duration` if `t` is
```
Typo: `ii` -> `is`

piodul commented on seastar/src/clocks.rs:
```diff
@@ -6,3 +8,134 @@ mod ffi {
 
 use ffi::*;
+
+// Macro used to generate the implemenetation of the instant types.
```
Typo: `implemenetation` -> `implementation`

piodul commented on seastar/src/clocks.rs:
```diff
@@ -6,3 +8,134 @@ mod ffi {
+            pub(crate) nanos: i64,
+        }
+
+        impl $Instant {
```
I think that most of the Instant's methods could be made const, right?

patjed41 replied:
I think all of them can be const and I changed all of them to const.

piodul commented on seastar/src/sleep.cc:
```diff
@@ -3,5 +3,13 @@
+
+VoidFuture lowres_sleep(int64_t nanos) {
+    co_await seastar::sleep<seastar::lowres_clock>(std::chrono::nanoseconds(nanos));
+}
```
Is there a reason why you didn't add a manual clock sleep?

patjed41 replied:
For some unknown reason I though `seastar::sleep` supports only steady and manual clocks. Of course, manual clock sleep can be added and I did that.

There is a slight problem with testing `manual_sleep`. I would like to spawn a task using `seastar::spawn` that advances `ManualClock` in a loop and then call and await `manual_sleep`. Unfortunately, I can't await `seastar::spawn` here and without awaiting `seastar::spawn` currently doesn't have any effect. I can write a complete test after someone fixes that. That's why there is only a very simple smoke test of `manual_sleep` now.

Edit: `seastar::spawn` was fixed and I implemented a complete test for `manual_sleep`.

patjed41 commented:
Update – generic `Duration`, `Instant` and `Sleep`:
- All clocks implement the `Clock` trait now. It contains only a `now` function.
- There is also a hidden `ClockImpl` trait that is a supertrait of `Clock`. All clock specific ffi functions should be put in this trait to avoid exposition outside the crate. There is only the `sleep` function for now. If we decide to stick to this solution, all ffi functions used by `Timer` will also go there.
- `ffi` module of the `timer` module was moved to the `clocks` module because it is used there.
- The code of `Duration`, `Instant` and `Sleep` is fully generic now.

I think current API has no significant disadvantages from the user's point of view. The code is fully generic, everything that should be public is public and everything that should be private is private. However, there is a small issue regarding the internal structure of the project. The `ClockImpl` trait forces part of the other modules' ffi (in this case its `sleep::ffi`) to be inside the `clocks` module which doesn't seem perfect. Keeping sleep's ffi in the `sleep` module and then `pub(crate)` exporting it would work here. However, problems would occur in the future if we did the same thing with timer's ffi. I can elaborate on that if we really want to avoid putting sleep's and timer's ffi in the `clocks` module.

Another issue with current changes is that putting
```rust
#[derive(Eq, PartialEq, Ord, PartialOrd, Hash, Clone, Copy, Default, Debug)]
```
before declaration of `Duration` and `Instant` doesn't work because of the `_phantom` field. I implemented all these traits manually. It could be shortened a little with the `derivative` crate but at the same time we would get another crate dependency and more complicated code.
